File: 191 Prize Strings.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# Date: 17 October 2017
# https://github.com/trizen

# Runtime: ~7 minutes.

# https://projecteuler.net/problem=191

# For a faster solution, see version 2.

function p_191(days::Int64)

    count = 0

    function generate(solution, hasL)

        len = length(solution)

        if (len == days)
            count += 1
        end

        if (len < days)

            if (len >= 2)
                if (solution[len] != 'A' || solution[len-1] != 'A')
                    generate(solution * "A", hasL)
                end
            else
                generate(solution * "A", hasL)
            end

            generate(solution * "O", hasL)

            if (!hasL)
                generate(solution * "L", true)
            end
        end
    end

    generate("", false)
    return count
end

println(p_191(30))

File: Perl/191 Prize Strings.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# Date: 19 October 2017
# https://github.com/trizen

# Runtime: 0.019s

# https://projecteuler.net/problem=191

use 5.022;
use strict;
use warnings;

use experimental qw(signatures);

sub p_191($days) {

    my %cache;
    sub ($len, $hasL, $hasA1, $hasA2) {

        exists($cache{"@_"})
            and return $cache{"@_"};

        my $count = ($len == $days) ? 1 : 0;

        if ($len < $days) {

            if (!$hasA1 || !$hasA2) {
                $count += __SUB__->($len + 1, $hasL, $hasA2, 1);
            }

            $count += __SUB__->($len + 1, $hasL, $hasA2, 0);

            if (!$hasL) {
                $count += __SUB__->($len + 1, 1, $hasA2, 0);
            }
        }

        return ($cache{"@_"} = $count);
    }->(0, 0, 0, 0);
}

say p_191(30);
